### DISCUSSION TITLE
Despite calling shutdown, getting "Fatal error: Client not shut down before the deinit"

### DISCUSSION THREAD

**Discussion by JetForMe** [soto-project/soto#728] [Q&A]
Soto v7.0.0:

I have code like this:

```swift
	func
	check(asset inAsset: Asset)
		async
		throws
	{
		let s3 = try createS3()
		
		do
		{
			guard let serverPath = inAsset.serverPath else { … }
			
			var logger = Logger(label: "S3Logger")
			logger.logLevel = .debug
			
			let req = S3.HeadObjectRequest(bucket: "", key: serverPath)
			let resp = try await s3.headObject(req, logger: logger)
			print("Head: \(resp)")
			
			try await s3.client.shutdown()
		}
		
		catch
		{
			try await s3.client.shutdown()
			throw error
		}
	}
```

Note that it calls `s3.client.shutdown()` whether or not it encounters an error.

But at the very end of the function, I get "AsyncHTTPClient/HTTPClient.swift:166: Fatal error: Client not shut down before the deinit. Please call client.shutdown() when no longer needed. Otherwise memory will leak."

The S3 service is created like this:

```swift
	func
	createS3()
		throws
		-> S3
	{
		guard
			let accessKey = self.accessKey,
			let secretKey = self.secretKey,
			let endpoint = self.endpoint?.absoluteString
		else
		{
			throw Errors.mediaServerNotConfigured
		}
		
		var config = HTTPClient.Configuration()
		config.httpVersion = .http1Only
		let logger = Logger(label: "S3Logger")

		let client = AWSClient(
			credentialProvider: .static(accessKeyId: accessKey, secretAccessKey: secretKey),
			middleware: AWSLoggingMiddleware(logger: logger, logLevel: .info),
			httpClient: HTTPClient(eventLoopGroupProvider: .singleton, configuration: config),
			logger: logger
		)
		let s3 = S3(client: client, region: .other(""), endpoint: endpoint, timeout: .hours(10))
		return s3
	}
```

Am I missing something?

**Update:** As I look through the code, I see it is `AsyncHTTPClient/HTTPClient` that is asserting it is not shut down, not `AWSClient`. Does `AWSClient.shutdown()` need to call `self.httpClient.shutdown()`? Or do I have to shut down both clients myself?

The more I look at it, the more of a mess it seems, as by default, `AWSClient` uses `HTTPClient.shared`, which probably shouldn't be shut down more than once. But in my app, there can be more than one S3 service in use, and they might need different HTTPClient configurations (the only reason I can’t use the shared one right now is because Digital Ocean Spaces seems to trip on some HTTP/2 requests).

**Comment by adam-fowler** [comment]
You created a separate HTTPClient that isn't managed by the AWSClient you have to shut that down as well.

**Reply by JetForMe** [reply]
Yeah, ok. These things really can't clean up after themselves, huh?
